assertj-core/src/main/java/org/assertj/core/api/AbstractIteratorAssert.java

```diff
@@ -93,4 +96,49 @@ public IterableAssert<ELEMENT> toIterable() {
     return new IterableAssert<>(IterableAssert.toIterable(actual));
   }
 
+  /**
+   * Verifies that the actual iterator is unmodifiable, i.e., throws an {@link UnsupportedOperationException} with
+   * any attempt to remove from the iterator.
+   * <p>
+   * Example:
+   * <pre><code class='java'> // assertions will pass
+   * assertThat(List.of().iterator()).isUnmodifiable();
+   * assertThat(Set.of().iterator()).isUnmodifiable();
+   *
+   * // assertions will fail
+   * assertThat(new ArrayList&lt;&gt;().iterator()).isUnmodifiable();
+   * assertThat(new HashSet&lt;&gt;().iterator()).isUnmodifiable();</code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual iterator is modifiable.
+   */
+  @Beta
+  public SELF isUnmodifiable() {
+    isNotNull();
+    assertIsUnmodifiable();
+    return myself;
+  }
+
+  private void assertIsUnmodifiable() {
+    switch (actual.getClass().getName()) {
+      case "java.util.Collections$EmptyIterator":
+      case "java.util.Collections$EmptyListIterator":
+        // immutable by contract, although not all methods throw UnsupportedOperationException
+        return;
+    }
```

Comment:
> To be short, I can't rely on that `iterator.remove()` will throw `UnsupportedOperationException`.
> 
> I.e. `Collections.EmptyIterator.remove()` throws `IllegalStateException`. Do we treat `EmptyIterator` as unmodifiable?

As you've already figured out, I don't think we have a better choice for these JDK types, similar to `emptyList`, `emptySet`, etc.

Reply:
Maybe I miss smth, but I can't find other iterator classes, only these two. Other collections create an interator "ad hoc".